Use dynamo fake tensor mode in aot_autograd, move aot_autograd compilation to lowering time [Merger of 89672 and 89773]

voznesenskym opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90089
* __->__ #90039

After all of the preparatory commits, this is a subset of the
changes in https://github.com/pytorch/pytorch/pull/89392 that actually
change us to propagating fake tensors to backends.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

This is the merger of Ed's PR #89672, which is a rewrite of an older PR of mine (#89392), with CI Fixes on top of it (#89773)

cc @mlazos @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

voznesenskym committed:
[UPDATED PROTOTYPE] Use dynamo fake tensor mode in aot_autograd, move aot_autograd compilation to lowering time

After all of the preparatory commits, this is a subset of the
changes in https://github.com/pytorch/pytorch/pull/89392 that actually
change us to propagating fake tensors to backends.

Signed-off-by: Edward Z. Yang <ezyangfb.com>


[WIP] Commit to run CI on vaguely sus C++ changes uncovered during fixing upstream prototype


Wip

Random stuff

Fix tests

Fix char

make shape prop fake tensor friendly

Fix

Test fixes

Feedback, test fixes, xla shenanigins

xla maddness

xla maddness

Feedback

undo

undo

Fix test

rm stupid stuff

[ghstack-poisoned]

ezyang commented on test/dynamo/test_dynamic_shapes.py:
```diff
@@ -66,6 +66,7 @@ def make_dynamic_cls(cls):
     # Cannot call sizes() on tensor with symbolic sizes/strides
 )
 
+
```
WHITESPAAAACE

voznesenskym replied:
k

ezyang commented on test/dynamo/test_modules.py:
```diff
@@ -703,6 +704,9 @@ def test_generation_tag(self):
 
+    # torch._subclasses.fake_tensor.UnsupportedFakeTensorException: meta converter nyi
+    # due to custom subclass (TensorProxy)
+    @unittest.expectedFailure
```
I have a fix for this in a local patchset, this xfail AOK

voznesenskym replied:
k

ezyang commented on test/dynamo/test_verify_correctness.py:
```diff
@@ -100,8 +100,11 @@ def compiler_fn(graph, example_inputs):
 
         self.assertIsNotNone(r1)
-        self.assertTrue(same(r1, r2))
-        self.assertTrue(same(r1, r3))
```
Why don't you just assert `same(r2, r3)` and call it a day lol

voznesenskym replied:
we care about r1 as well

ezyang commented on torch/fx/passes/shape_prop.py:
```diff
@@ -109,11 +109,36 @@ def forward(self, x):
+                self.module = self.fake_module
+
             result = super().run_node(n)
+            self.module = self.real_module
```
If run_node raises an exception this won't reset the module.

ezyang replied:
There's no need to do a swap like this. The only thing you need to override is a few ops (`get_attr`, `call_module` I think?) to fetch from fake module rather than real module.

voznesenskym replied:
good call

voznesenskym replied:
> If run_node raises an exception this won't reset the module.

good call

> There's no need to do a swap like this. The only thing you need to override is a few ops (`get_attr`, `call_module` I think?) to fetch from fake module rather than real module.

Yeah but we already override this, this feels simpler. No strong opinion tho.

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -398,6 +398,10 @@ def clone_tensor(x):
+    # TODO: this is questionable
+    if isinstance(x, torch._subclasses.FakeTensor):
+        # this func fails on fake tensors in __torch_dispatch__
+        return x
```
I'll put up with this because I think my pending refactor will resolve the confusion here

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -1155,3 +1165,20 @@ def assert_no_fake_params_or_buffers(gm):
+            if fake_mode is None:
+                fake_mode = flat_input.fake_mode
+            else:
+                assert fake_mode == flat_input.fake_mode
```
is, not equality, plz

ezyang commented on torch/_dynamo/optimizations/backends.py:
```diff
@@ -747,9 +747,16 @@ def torchxla_trivial(subgraph):
+
+    def fwd(*args):
+        nonlocal compiled_graph
+        model = subgraph.model
```
nit: this will keep subgraph permanently alive, whereas previously it could have been GC'ed after compilation. You should del subgraph being done with compilation

voznesenskym replied:
good idea

ezyang commented on torch/_dynamo/optimizations/distributed.py:
```diff
@@ -211,7 +214,7 @@ def __init__(self, module, compiler):
                 self.compiler = compiler
 
-            def compile_submod(self, submod, args, kwargs):
+            def compile_submod(self, input_mod, args, kwargs):
```
The variable renaming makes it harder for code reviewers

ezyang replied:
like, afaict, there's literally no change from lines 211 to 250 but I had to carefully audit to make sure there weren't typos lol

voznesenskym replied:
This is from your PR?

ezyang replied:
which is originally from https://github.com/pytorch/pytorch/pull/89392/files#diff-ebfc1054a8736abfc050c0a51c1f64f323a8cc3088f3f621d628cf7b421401faR219 lol

ezyang commented on torch/_dynamo/optimizations/distributed.py:
```diff
@@ -236,42 +239,61 @@ def forward(self, *args):
 
             def run_node(self, n: Node) -> Any:
                 with fx_traceback.append_stack_trace(n.stack_trace):
                     args, kwargs = self.fetch_args_kwargs_from_env(n)
```
It would be really nice to have a comment here explaining what's going on

voznesenskym replied:
Sure thing

ezyang commented on torch/_dynamo/optimizations/distributed.py:
```diff
@@ -236,42 +239,61 @@ def forward(self, *args):
+                        if isinstance(arg, torch.Tensor) and not isinstance(
+                            arg, torch._subclasses.FakeTensor
+                        ):
+                            new_args.append(fake_mode.from_tensor(arg))
```
It would be much better if we could assume all the args are already fakeified (and we just maintain the invariant that all the intermediates are fake tensors). In particular, fakeifying a tensor can trigger the allocation of new symints, but if they're not properly associated as inputs we may not be able to instantiate those variables on subsequent runs.

tl;dr I suspect arg is always a fake tensor here

voznesenskym replied:
Im not sure we can, I'm pretty sure we saw non fakes in here, but maybe that was in an intermediate state of this PR. I can check again.

ezyang replied:
you can check this by pushing a separate PR that tightens the invariant and see if it fails or not

ezyang commented on torch/_dynamo/optimizations/distributed.py:
```diff
@@ -236,42 +239,61 @@ def forward(self, *args):
+                            fake_submod = deepcopy_to_fake_tensor(real_mod, fake_mode)
+                        else:
+                            fake_submod = real_mod
+                            pass
```
pass here unnecessary

ezyang commented on torch/_dynamo/output_graph.py:
```diff
@@ -88,21 +88,10 @@ def __repr__(self):
-
 class WrapperBackend:
-    def __init__(self, backend: CompilerFn):
+    def __init__(self, backend, original_example_inputs):
```
y u delete the type annotation lol

ezyang commented on torch/_dynamo/output_graph.py:
```diff
@@ -541,16 +529,42 @@ def call_user_compiler(self, gm: fx.GraphModule) -> CompiledFn:
+                result.extend(example)
+            else:
+                # Fallback, in case fake_tensor was not set
+                # Particularly for graph args that are not tensors
```
I hope graph args can't be None, because otherwise there is type confusion here.

ezyang commented on torch/_dynamo/output_graph.py:
```diff
@@ -541,16 +529,42 @@ def call_user_compiler(self, gm: fx.GraphModule) -> CompiledFn:
+            # We invoke the backend with real inputs if and only if
+            # we are in doing accuracy evaluation in the minifier
+            # Note: This check is the same as in `save_graph_repro`
+            is_minifier_backend = "_accuracy" in str(
```
nuh uh, we're not testing for minifier backend this way. Why can't you just do an `is` test here, there's what, one function you need to test?

ezyang commented on torch/_dynamo/output_graph.py:
```diff
@@ -541,16 +529,42 @@ def call_user_compiler(self, gm: fx.GraphModule) -> CompiledFn:
+            is_minifier_backend = "_accuracy" in str(
+                torch._dynamo.eval_frame.innermost_fn(compiler_fn)
+            )
+            is_top_level_minifiying = (
```
nit: minifying

ezyang commented on torch/_dynamo/variables/builder.py:
```diff
@@ -108,8 +108,13 @@ class GraphArg:
     def __post_init__(self):
+        if isinstance(self.example, torch.Tensor):
+            assert self.fake_tensor is not None and isinstance(
+                self.fake_tensor, torch._subclasses.fake_tensor.FakeTensor
```
you don't need to do the is None test, with the isinstance test

ezyang commented on torch/_dynamo/variables/builder.py:
```diff
@@ -689,11 +713,14 @@ def _clone_input(value):
-            fake_wrapper = functools.partial(wrap_to_fake_tensor_and_record, tx=tx)
-            example_value = fake_wrapper(example_value)
+            # Note: Unfortunately, this can happen during tracing, and is valid enough for now to allow.
+            # TODO(voz): Find all the callsites and burn this down.
```
I think I also managed to nail a bunch of these in pending PR

ezyang commented on torch/_dynamo/variables/builder.py:
```diff
@@ -551,10 +565,7 @@ def wrap_tensor(self, value: torch.Tensor):
-                self.tx.output.graphargs.append(
-                    GraphArg(self.get_source(), value, False)
-                )
+            graph_arg = None
```
this None assignment is unnecessary

ezyang commented on test/dynamo/test_optimizations.py:
```diff
@@ -120,6 +120,38 @@ def compiler_fn(graph, example_inputs):
+
+        a = torch.empty(2).fill_(1)
+        b = torch.empty(2).fill_(2)
+        c = torch.empty(2).fill_(3)
```
The code duplication here is a bit irritating. Is there something in particular you are leaving `test_example_inputs` around to test?

voznesenskym replied:
A backend that returns straight from example input should have fakes but be correct.

ezyang commented on torch/_dynamo/test_minifier_common.py:
```diff
@@ -100,13 +99,14 @@ def _run_repro(self, patch_code, repro_dir):
     # `patch_code` is the code to be patched in every generated file.
-    def _gen_test_code(self, run_code, repro_after, repro_level, patch_code):
+    def _gen_test_code(
+        self, run_code, repro_after, repro_level, patch_code, fake_tensor
```
Do you even use `fake_tensor`?

voznesenskym committed:
Update on "Use dynamo fake tensor mode in aot_autograd, move aot_autograd compilation to lowering time [Merger of 89672 and 89773]"


After all of the preparatory commits, this is a subset of the
changes in https://github.com/pytorch/pytorch/pull/89392 that actually
change us to propagating fake tensors to backends.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

This is the merger of Ed's PR #89672, which is a rewrite of an older PR of mine (#89392), with CI Fixes on top of it (#89773)

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Use dynamo fake tensor mode in aot_autograd, move aot_autograd compilation to lowering time [Merger of 89672 and 89773]"


After all of the preparatory commits, this is a subset of the
changes in https://github.com/pytorch/pytorch/pull/89392 that actually
change us to propagating fake tensors to backends.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

This is the merger of Ed's PR #89672, which is a rewrite of an older PR of mine (#89392), with CI Fixes on top of it (#89773)

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang commented on torch/_dynamo/debug_utils.py:
```diff
@@ -762,6 +762,8 @@ def backend_accuracy_fails(gm, example_inputs, compiler_fn, only_fwd=False):
 
 backend_aot_accuracy_fails = functools.partial(backend_accuracy_fails, only_fwd=True)
 
+MINIFIER_SPAWNED = False
```
Please add comment here, explaining why this exists, how it is used, and when we would be able to remove it

ezyang commented on torch/_dynamo/optimizations/distributed.py:
```diff
@@ -236,42 +239,75 @@ def forward(self, *args):
 
+            # Note:
+            #
+            # The way distributed works today we fakify the module for compile time,
```
The confusion here is what is meant by "compile time", which can be reasonably miscontrued to mean "what I pass to the compiler function" whereas what you actually mean is "compile time fake tensor propagation, so we can pass fake tensors into the submodule compiles"

voznesenskym replied:
I'll rewrite it.

ezyang commented on torch/_dynamo/output_graph.py:
```diff
@@ -541,16 +529,38 @@ def call_user_compiler(self, gm: fx.GraphModule) -> CompiledFn:
+                compiler_fn = WrapperBackend(compiler_fn, self.example_inputs())
+
+            # We invoke the backend with real inputs if and only if
+            # we are in doing accuracy evaluation in the minifier
```
Comment for why this is a hack and when it can be removed

voznesenskym committed:
Update on "Use dynamo fake tensor mode in aot_autograd, move aot_autograd compilation to lowering time [Merger of 89672 and 89773]"


After all of the preparatory commits, this is a subset of the
changes in https://github.com/pytorch/pytorch/pull/89392 that actually
change us to propagating fake tensors to backends.

Signed-off-by: Edward Z. Yang <ezyangfb.com>

This is the merger of Ed's PR #89672, which is a rewrite of an older PR of mine (#89392), with CI Fixes on top of it (#89773)

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

voznesenskym committed:
Update backends.py

voznesenskym commented:
@pytorchbot merge

voznesenskym committed:
Update backends.py

voznesenskym commented:
@pytorchbot merge -f "weird unrelated py3.7 pip install bug in manywheel"

clee2000 commented:
@pytorchbot revert -m "broke xla tests https://hud.pytorch.org/pytorch/pytorch/commit/ef0c7ec958439caf44a98fb7b70d920c6c2264b9 https://github.com/pytorch/pytorch/actions/runs/3606308473/jobs/6077646142" -c landrace

voznesenskym commented:
> @pytorchbot revert -m "broke xla tests https://hud.pytorch.org/pytorch/pytorch/commit/ef0c7ec958439caf44a98fb7b70d920c6c2264b9 https://github.com/pytorch/pytorch/actions/runs/3606308473/jobs/6077646142" -c landrace

I'd rather we have forward fixed, its a 1 liner to fix. Alas.

voznesenskym committed:
Fix XLA

ezyang commented:
the way to correctly land xla changes is to update the hash in the PR to your branch on xla. Once landed, merge the xla branch into master. The automatic head updating process will eventually reset xla's commit id to master.

voznesenskym commented:
> the way to correctly land xla changes is to update the hash in the PR to your branch on xla. Once landed, merge the xla branch into master. The automatic head updating process will eventually reset xla's commit id to master.

Yes. But I did it in a way that should not need XLA changes.

voznesenskym commented:
@pytorchbot merge -f "weird unrelated py3.7 pip install bug in manywheel"

voznesenskym commented:
@pytorchmergebot / @pytorchbot merged the wrong commit >:(
